### PR TITLE
Add basline tests

### PR DESCRIPTION
**Pull request by SchwammDev**
These tests make sure, that a baseline appropriately reduces the variance and helps the algorithm to learn a good policy in a reasonable amount of time. Tests have been devises which are especially hard when they are not corrected by a baseline. For instance it is very hard for reinforce to find a good reward signal in the one dimensional walk, because of the high variance and back propagation of the reward.

### CODE REVIEW

File: reinforcement/algorithm/reinforce.py [hunk 1]
```diff
@@ -19,16 +23,21 @@ def sample(self, observation):
 
     def optimize(self, trajectory):
         self._update_rewards_to_return_signals(trajectory)
-        self._policy.fit(trajectory)
+        self._trajectories.append(trajectory)
+        if len(self._trajectories) == self._num_train_trajectories:
+            total = sum(self._trajectories, Trajectory())
```
Comment (SchwammDev) [inline reinforcement/algorithm/reinforce.py]: join/chain/concatenate would probably be clearer as python like function instead of sum

File: reinforcement/algorithm/reinforce.py [hunk 2]
```diff
@@ -19,16 +23,21 @@ def sample(self, observation):
...
 
     def _update_rewards_to_return_signals(self, trajectory):
-        bases = self._baseline.estimate(trajectory)
         trj_len = len(trajectory)
         for tp in range(trj_len, 0, -1):
             prev = trajectory.returns[tp] if tp < trj_len else 0
             trajectory.returns[tp - 1] += self._gamma * prev
-        trajectory.returns -= bases
-        self._normalize(trajectory.returns)
+        bases = self._baseline.estimate(trajectory)
```
Comment (SchwammDev) [inline reinforcement/algorithm/reinforce.py]: could inline

File: reinforcement/trajectories.py [hunk 3]
```diff
@@ -15,6 +17,16 @@ def __eq__(self, other):
                np.array_equal(self.observations, other.observations) and \
                np.array_equal(self.returns, other.returns)
 
+    def __add__(self, other):
```
Comment (SchwammDev) [inline reinforcement/trajectories.py]: better done with some join/chain function - see above

File: tests/aux/baselines.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+        self._summary_writer = summary_writer
+        logs = [tf1.summary.scalar("mean_baseline_return", tf.reduce_mean(self._in_returns)),
+                tf1.summary.scalar("baseline_loss", loss),
+                _log_2d_tensor_as_img("baseline_nn", nn)]
+        self._log_summary = tf1.summary.merge(logs)
+        self._cur_episode = 0
+
+    def estimate(self, trj):
+        r = trj.returns - np.ones_like(trj.returns) * 0.1
```
Comment (SchwammDev) [inline tests/aux/baselines.py]: This seems like a leftover from testing - delete

File: tests/aux/baselines.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        return self._session.run(self._out_prediction,
+                                 {self._in_observations: trj.observations,
+                                  self._in_returns: r})
+
+    def fit(self, trj):
+        r = trj.returns - np.ones_like(trj.returns) * 0.1
```
Comment (SchwammDev) [inline tests/aux/baselines.py]: This seems like a leftover from testing - delete

File: tests/aux/policies.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+        def show(self):
+            pass
+
+
+def _log_2d_tensor_as_img(name, mat):
+    return tf1.summary.image(name, tf.reshape(mat, shape=(1, mat.shape[0].value, mat.shape[1].value, 1)))
+
+
+INDEX = 0
```
Comment (SchwammDev) [inline tests/aux/policies.py]: Seems to be a testing leftover - delete

File: tests/aux/policies.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+
+class ParameterizedPolicy:
+    def __init__(self, session, obs_dims, num_actions, summary_writer, lr=10):
+        global INDEX
```
Comment (SchwammDev) [inline tests/aux/policies.py]: Seems to be a testing leftover - delete

File: tests/aux/policies.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+                                 initializer=tf.glorot_uniform_initializer())
+        self._out_probabilities = tf.nn.softmax(tf.matmul(self._in_observations, theta))
+        self._train = None
+
+        self._logs = [tf1.summary.scalar(f"mean_normalized_return_{INDEX}", tf.reduce_mean(self._in_returns)),
+                      _log_2d_tensor_as_img(f"theta_{INDEX}", theta)]
+        self._log_summary = tf.no_op
+        self._summary_writer = summary_writer
+        self._cur_episode = 0
+        INDEX += 1
```
Comment (SchwammDev) [inline tests/aux/policies.py]: Seems to be a testing leftover - delete

File: tests/aux/policies.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+class ParameterizedPolicy:
+    def __init__(self, session, obs_dims, num_actions, summary_writer, lr=10):
+        global INDEX
+        self._session = session
+        self._lr = lr
+        self._in_actions = tf1.placeholder(shape=(None,), dtype=tf.uint8, name="actions")
+        self._in_returns = tf1.placeholder(shape=(None,), dtype=tf.float32, name="returns")
+        self._in_observations = tf1.placeholder(shape=(None, obs_dims), dtype=tf.float32, name="observations")
+        theta = tf1.get_variable(f"theta_{INDEX}", shape=(obs_dims, num_actions), dtype=tf.float32,
```
Comment (SchwammDev) [inline tests/aux/policies.py]: Seems to be a testing leftover - delete

File: tests/aux/policies.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+                                 initializer=tf.glorot_uniform_initializer())
+        self._out_probabilities = tf.nn.softmax(tf.matmul(self._in_observations, theta))
+        self._train = None
+
+        self._logs = [tf1.summary.scalar(f"mean_normalized_return_{INDEX}", tf.reduce_mean(self._in_returns)),
```
Comment (SchwammDev) [inline tests/aux/policies.py]: Seems to be a testing leftover - delete

File: tests/conftest.py [hunk 11]
```diff
@@ -2,6 +2,8 @@
 
 import numpy as np
 import pytest
+import tensorflow as tf
+from tensorflow._api.v1.compat import v1 as tf1
```
Comment (SchwammDev) [inline tests/conftest.py]: This seems strange is there not an import ... version -- also import error is needed

File: tests/aux/policies.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
+import numpy as np
+import tensorflow as tf
+from tensorflow._api.v1.compat import v1 as tf1
```
Comment (SchwammDev) [inline tests/aux/policies.py]: Import error needed also strange v1 import

File: tests/aux/baselines.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
+import numpy as np
+import tensorflow as tf
+from tensorflow._api.v1.compat import v1 as tf1
```
Comment (SchwammDev) [inline tests/aux/baselines.py]: Import error needed also strange v1 import

File: tests/test_reinforce_algorithm.py [hunk 14]
```diff
@@ -5,20 +5,19 @@
...
+        self.fit_onto = None
         self._calc = None
 
     def set_signal_calc(self, calc):
         self._calc = calc
 
     def fit(self, trajectory):
-        CALL_ORDER.append(self.fit)
-        self.received_returns = list(trajectory.returns)
+        self.received_advantages = trajectory.advantages
```
Comment (SchwammDev) [inline tests/test_reinforce_algorithm.py]: Ambiguity! Use just one

File: tests/test_reinforce_algorithm.py [hunk 15]
```diff
@@ -55,21 +54,28 @@ def __init__(self, size):
...
+    def estimate(self, trj):
+        self.received_returns = trj.returns
+        return super().estimate(trj)
 
-@pytest.fixture(autouse=True)
-def cleanup_call_order():
-    CALL_ORDER.clear()
+    def fit(self, trj):
+        super().fit(trj)
+        self.fit_onto = trj.returns
```
Comment (SchwammDev) [inline tests/test_reinforce_algorithm.py]: This for sure will lead to confusion - in the Spy class, this is set to advantages, while here it is overridden with returns.

File: tests/test_reinforce_algorithm.py [hunk 16]
```diff
@@ -92,41 +98,77 @@ def test_sampling_returns_estimate_of_action_probabilities(policy, observation):
...
-    alg = make_alg(policy, baseline)
-    alg.optimize(trajectory)
-    assert CALL_ORDER == [baseline.estimate, policy.fit]
+def test_that_reinforce_passes_accumulated_returns_to_baseline(policy, baseline, make_trajectory):
+    alg = make_alg(policy, baseline, gamma=0.5)
+    alg.optimize(make_trajectory(returns=[1.0, 2.0, 3.0]))
+    assert_estimates(baseline.received_returns, [2.75, 3.5, 3.0])
+
+
+def test_reinforce_does_not_fit_policy_and_baseline_until_trajectory_batch_is_full(policy, baseline, make_trajectory):
```
Comment (SchwammDev) [inline tests/test_reinforce_algorithm.py]: ...is_filled(...

File: tests/test_reinforce_algorithm.py [hunk 17]
```diff
@@ -5,20 +5,19 @@
...
         self._calc = None
 
     def set_signal_calc(self, calc):
         self._calc = calc
 
     def fit(self, trajectory):
-        CALL_ORDER.append(self.fit)
-        self.received_returns = list(trajectory.returns)
+        self.received_advantages = trajectory.advantages
+        self.fit_onto = self.received_advantages
```
Comment (SchwammDev) [inline tests/test_reinforce_algorithm.py]: If I keep this one fitted_onto might be better

File: tests/test_reinforce_algorithm.py [hunk 18]
```diff
@@ -92,41 +98,77 @@ def test_sampling_returns_estimate_of_action_probabilities(policy, observation):
...
+    baseline.set_estimates([2.0, 3.0, 4.0])
+    alg.optimize(make_trajectory(returns=[1.0, 2.0, 3.0]))
+    alg.optimize(make_trajectory(returns=[4.0, 3.0, 2.0]))
+    assert_signals(policy.fit_onto, concat(normalized([0.75, 0.5, -1.0]), normalized([4.0, 1.0, -2.0])))
+
+
+def test_reinforce_uses_clean_trajectory_batches(policy, baseline, make_trajectory):
+    baseline.set_estimates([0, 0, 0])
+    alg = make_alg(policy, baseline, gamma=0.5, num_train_trjaectories=2)
+    for _ in range(2):
```
Comment (SchwammDev) [inline tests/test_reinforce_algorithm.py]: put this into a descriptive "fill_batch" or similar function

File: tests/test_reinforce_algorithm.py (hunk 15 above)
Comment (SchwammDev) [reply tests/test_reinforce_algorithm.py]: Nope, this is the baseline spy so it works as expected
